blog/you-cant-create-a-constant-using-usememo-in-react.md

```diff
@@ -0,0 +1,58 @@
+# You can't create a constant using useMemo in React
+
+There are 3 possible options to create a constant in React: `useState`, `useMemo` and `useRef`. We're going to see which one is the best way to create real constants (not constant-like values).
+
+I talked about the [`useConst`](https://github.com/chakra-ui/chakra-ui/blob/main/packages/hooks/src/use-const.ts) hook from [chakra-ui](https://chakra-ui.com) in a [previous post](https://dev.to/cloudx/chakra-ui-the-hidden-treasures-1gl9), and now, I found the same hook in [Fluent UI](https://www.npmjs.com/package/@fluentui/react-hooks#useconst), we are going to review why they've implemented it.
+
+## Remember useMemo
+
+The first thought that comes to mind when we look for a constant in a render function is to use the [`useMemo`](https://reactjs.org/docs/hooks-reference.html#usememo) hook. This is a good option, but only if your "constant" value shouldn't change based on the dependencies.
+
+```javascript
+const memoizedValue = useMemo(
+  () => computeExpensiveValue(a, b),
+  [a, b]
+);
+```
+
+There is no warranty that the returned value of `useMemo` is always the same even when the dependencies didn't change. React can run the factory function if the tool believes that the value should be re created.
+
+🧠 The official documentation says:
+
+> Write your code so that it still works without `useMemo` — and then add it to optimize performance.
+
+## The useState option
+
+If a state didn't change, then it's a constant.
+
+```javascript
+const [value] = useState(initialValue);
+```
+
+Yeah, we can create a constant with [`useState`](https://reactjs.org/docs/hooks-reference.html#usestate) but it's expensive due to reducer handling which we don't need.
+
+🧙‍♂️ If you want to learn how React handles its hooks inside the Fiber reconciler, you should read the [Fiber hooks](https://github.com/facebook/react/blob/main/packages/react-reconciler/src/ReactFiberHooks.new.js) code.
+
+## A useRef approach
+
+The [`useRef`](https://reactjs.org/docs/hooks-reference.html#useref) hook can hold a value, and doesn't have an internal reducer or checks on the dependencies array, also React doesn't re-create the value for performance reasons.
+
+```javascript
+const useConst = (initialValue) => {
+  const ref = React.useRef();
+  if (ref.current === undefined) {
+    ref.current = typeof initialValue === 'function' ? initialValue() : initialValue,
```

Comment:
Is the `,` right at the end of this line?

Reply:
```suggestion
    ref.current = typeof initialValue === 'function' ? initialValue() : initialValue;
```